### PR TITLE
Pull data monthly in fact sheet

### PR DESCRIPTION
**Pull request by Rohit25negi**
This is from the Latest learning that Indexes on month column are not used when we pull multiple months data via `in` or `BETWEEN` or `< and >` queries. They are only used in `=` query. Last we have seen this in the child growth tracker report that pulling month data separately for each month is much faster than pulling all of the months data together because of the index thing mentioned above.

I am positive that this would reduce the load time of Factsheet a lot. I also think that this will prevent the occasional high spikes in the load time which I guess happens because of the super slow queries Will be testing it out before actually merging it. Putting it out as Pr so to get the review before do the limited release.

### CODE REVIEW

**Review by czue — APPROVED** [review]
code looks good. i don't have the knowledge to weigh in on the performance but sounds like you're confident

**Comment by Rohit25negi** [comment]
@calellowitz  could you take a look at this one please. I am planning to first test it via a limited release before merging

**Comment by calellowitz** [comment]
> @calellowitz could you take a look at this one please. I am planning to first test it via a limited release before merging

How does one test a database change in a limited release? You have either added the index to the DB or you haven't

**Comment by Rohit25negi** [comment]
> > @calellowitz could you take a look at this one please. I am planning to first test it via a limited release before merging
> 
> How does one test a database change in a limited release? You have either added the index to the DB or you haven't

hey, its. not a db change. I just changed how it fetches the data. @calellowitz

File: custom/icds_reports/reports/fact_sheets.py [hunk 1]
```diff
@@ -623,22 +612,35 @@ def get_data(self):
...
-        data_sources = [
-            data_source.get_data()
-            for k, data_source in self.data_sources(self.config).items()
-            if k in needed_data_sources
-        ]
-
-        all_data = self._get_all_data(data_sources)
-
-        sql_location = self.config.get('sql_location')
+        data_sources = list()
```
Comment (calellowitz) [inline custom/icds_reports/reports/fact_sheets.py:615]: nit, `data_sources = []` is more standard in our code base

File: custom/icds_reports/reports/fact_sheets.py [hunk 2]
```diff
@@ -623,22 +612,35 @@ def get_data(self):
...
         months = [
             dt.strftime("%b %Y") for dt in rrule(
                 MONTHLY,
                 dtstart=self.config['two_before'],
                 until=self.config['month'])
         ]
 
+        for data_source_name in needed_data_sources:
+            data_row = list()
```
Comment (calellowitz) [inline custom/icds_reports/reports/fact_sheets.py:624]: same nit

**Review by calellowitz — APPROVED** [review]
Looks fine. Would be good to include the profiling so we can see how this is better, since its not intuitive to me that it will be (its a lot more queries), but if it shows improvement then we can go forward. It might also  be worth looking into adding indices that can be used by a between query, if the issue is that our indices are insufficient, since that will allow us to keep down the number of queries

File: custom/icds_reports/reports/fact_sheets.py (hunk 1 above)
Comment (Rohit25negi) [reply custom/icds_reports/reports/fact_sheets.py:615]: I see. I just feel `list()` more pythonic :P But sure can make it `[]` after profiling is done

**Comment by Rohit25negi** [comment]
I tried profiling and here are the results and they look promising:

1. Overall time difference: I have pulled the data for 10 times in a loop and following are the overall time required by factsheet in each run:

Existing Code:

```
0:00:04.042751
0:00:01.302446
0:00:01.414177
0:00:01.301646
0:00:01.479661
0:00:01.773990
0:00:02.468092
0:00:01.272388
0:00:01.237815
0:00:01.260979

```


With this change

```
0:00:01.576954
0:00:02.102530
0:00:00.394208
0:00:00.415842
0:00:00.818125
0:00:00.848980
0:00:00.837384
0:00:00.625123
0:00:00.574757
0:00:00.798429

```



Below is the full profiling


Existing code:
[existing_code_profiling.txt](https://github.com/dimagi/commcare-hq/files/4869169/existing_code.txt)


```

cProfile.run('func()')                                                                                                                               
         75141 function calls (74232 primitive calls) in 1.402 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    1.402    1.402 <ipython-input-4-2d8ffe86448c>:1(func)
        3    0.000    0.000    0.006    0.002 <string>:1(<lambda>)
        1    0.000    0.000    1.402    1.402 <string>:1(<module>)
        3    0.000    0.000    0.000    0.000 <string>:1(__init__)
       52    0.000    0.000    0.001    0.000 <string>:1(_create_text)

```


With this change:
[with_new_code_profile.txt](https://github.com/dimagi/commcare-hq/files/4869173/with_new_code.txt)


```

cProfile.run('func()')                                                                                                                               
         118313 function calls (115703 primitive calls) in 0.627 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.627    0.627 <ipython-input-4-2d8ffe86448c>:1(func)
        9    0.000    0.000    0.014    0.002 <string>:1(<lambda>)
        1    0.000    0.000    0.627    0.627 <string>:1(<module>)
        9    0.000    0.000    0.000    0.000 <string>:1(__init__)
      156    0.000    0.000    0.002    0.000 <string>:1(_create_text)
        9    0.000    0.000    0.000    0.000 <string>:1(asc)
      126    0.000    0.000    0.001    0.000 <string>:1(bindparam)
       78    0.000    0.000    0.002    0.000 <string>:1(case)
      204    0.000    0.000    0.001    0.000 <string>:1(column)
        9    0.000    0.000    0.000    0.000 <string>:1(order_by)

```


@calellowitz  how does it look to you?

**Comment by Rohit25negi** [comment]
@calellowitz  humble ping on this. I am looking to deploy this today

**Pull request merged**